content/shmem_init_thread.tex

```diff
@@ -0,0 +1,78 @@
+\apisummary{
+Initializes the \openshmem{} library, similar to SHMEM\_INIT(), and in addition
+perform the initialization required for thread-safe invocation of \openshmem{} functions.}
+
+\begin{apidefinition}
+
+\begin{Csynopsis}
+int shmem_init_thread(int requested, int *provided);
+\end{Csynopsis}
+
+\begin{Fsynopsis}  
+INTEGER requested, provided
+     
+SHMEM_INIT_THREAD(requested, provided)
+\end{Fsynopsis}  
+
+\begin{apiarguments}
+\apiargument{IN}{requested}  {The thread level support requested by the user.
+The correct values are SHMEM\_THREAD\_SINGLE, SHMEM\_THREAD\_FUNNELED, or SHMEM\_THREAD\_MULTIPLE}
+\apiargument{OUT}{provided}{The thread level support provided by the \openshmem{} implementation.}
+\end{apiarguments}
+
+
+\apidescription{
+    \FUNC{shmem\_init\_thread}  initializes the \openshmem{} library similar to
+    \FUNC{shmem\_init}, and in addition perform the initialization required for
+    thread-safe invocation of \openshmem{} functions. The argument
+    \textit{requested} is used to specify the desired level of thread support.
+    The function returns the support provided by the library. There are four 
+    levels of threading support.
+ 
+ \begin{itemize}
+\item SHMEM\_THREAD\_SINGLE: The \openshmem{} program may not be multithreaded, 
+and all \openshmem{} interfaces are invoked by a single thread. 
+
+\item SHMEM\_THREAD\_FUNNELED: 
+The \openshmem{} program may be multithreaded, however, the 
+program must ensure that only one thread invokes the \openshmem{}
+interfaces.
+
+\item SHMEM\_THREAD\_SERIALIZED: 
+The \openshmem{} program may be multithreaded, however, the 
+program must ensure that only one thread invokes the \openshmem{}
+interfaces at any given instance i.e., the \openshmem{} interfaces 
+are not invoked concurrently by multiple threads.
+
+\item SHMEM\_THREAD\_MULTIPLE: 
+The \openshmem{} program may be multithreaded, and any 
+thread may invoke the \openshmem{} interfaces.
+
+\item SHMEM\_THREAD\_UNSUPPORTED: The \openshmem{} specification 
+can be implemented without the support for threads, or a model beyond the four
+models.
+   
+\end{itemize}
+
+The function may be used to initialize \openshmem{}, and to initialize the
+\openshmem{} with thread safety, instead of SHMEM\_INIT. The SHMEM\_INIT\_THREAD
+may not be called multiple times in an \openshmem{} program.
```

Comment:
> @naveen-rn As long as, the context can be created in SHARED or PRIVATE mode, and can be used with/without threads, it should be enough to support this. I’m not sure, I understand the new challenge.

As per my understanding from the current standing Context proposal, Contexts can be created either by explicitly mentioning the property like PRIVATE/SERIALIZED, or they can be created with a default property. The default property depends on the value of the *threading model set during initialization*. In this example;
```C
#define DEFAULT 0 /* If 0 is used on the context property - the default property is set */

shmem_init();
shmem_ctx_create(DEFAULT, ctx1); /* ctx1 property PRIVATE because of shmem_init */
shmem_ctx_create(SHARED, ctx2);

shmem_init_thread(SHMEM_THREAD_MULTIPLE);
shmem_ctx_create(DEFAULT, ctx3); /* ctx3 property SHARED because of shmem_init_thread */
``` 
1. The default property of the Contexts is inconsistent across different regions in the application
2. When the default property changes, should we modify the existing property of the already created Contexts(probably not - too expensive) to make it consistent?
3. And, then there is SHMEM_CTX_DEFAULT - this definitely needs modification of its existing property before and after shmem_init_thread(). I think this particular change is reasonable.

> Yes. The implementations can avoid all threading overheads until the shmem_init_thread is called. So, in your example, the first shmem_put might be implemented more efficiently than the second.

If removing threading overheads is the main reason behind allowing users to use the shmem_init(); shmem_init_thread() sequence, why don't they just use a PRIVATE context to achieve that?  

**PS:** After reading my own reply, I feel this seems to be more like a Context related discussion.  If you feel this usage sequence is very important, we could handle this in the Context proposal.